compiler/qsc_parse/src/item.rs

```diff
@@ -245,9 +245,27 @@ fn parse_callable_decl(s: &mut Scanner) -> Result<Box<CallableDecl>> {
     } else {
         Vec::new()
     };
+    // ensure that the function inputs are enclosed in parens.
+    // see https://github.com/microsoft/qsharp/issues/409 for context
+    // we check than an open parens precedes the input, but a check
+    // that a succeeding closed parens is unnecessary. This is because,
+    // if an open parens is present, then the `pat` function will only
+    // recognize a pattern with a matching closing parens.
+    let tok = s.peek();
+    match tok.kind {
+        TokenKind::Open(Delim::Paren) => (),
```

Comment:
While this works, my preference is to avoid low-level token reasoning when high-level tree reasoning is available. Because the issue is that the parser is too permissive, you have the benefit of an AST that you can check after-the-fact to make sure that the callable input pattern is a `PatKind::Paren`.

See [`stmt::check_semis`](https://github.com/microsoft/qsharp/blob/c35d6257c5eaa8c463df78104a421eb518d37c96/compiler/qsc_parse/src/stmt.rs#L166) for an example of this.